addons/web_editor/static/src/js/editor/snippets.editor.js

```diff
@@ -2389,6 +2389,14 @@ var SnippetsMenu = Widget.extend({
                     }
 
                     $toInsert = $baseBody.clone();
+                    // The associated snippet widget should be started so that
+                    // the user visualizes the final version of the snippet
+                    // that is being dragged.
+                    self.trigger_up('widgets_start_request', {
+                        $target: $toInsert,
+                        editableMode: true,
+                    });
```

Comment:
Mmh... really scared of that change, it can't be ok. No accounting of async operations, starting a widget on an element which is not in the body so, what the widget computes may be wrong, etc, ... :grimacing: 

I'll have to think about it... if I remember well last time I discussed about this with PDE / previous dev of the task... it was decided to use the "s_preview" system you speak about in your commit message to show an image of what will be dropped, I guess the spec changed again :grimacing:

Reply:
I left it as it's not possible to display a static template for the countdown and the charts previews, as the canvases are drawn by the widgets... Maybe we could display an image with s_preview for those cases ?

Reply:
Not sure if that's what you speak of, but back then when HRE was in charge of this task, I created and sent him transparent image version of the countdown that was supposed to be used as preview.
In case..
<details>
  <summary>See here!</summary>
  
![svgviewer-output](https://user-images.githubusercontent.com/30048408/147740260-aa010acb-b502-4dbd-8848-15f5f0e07425.png)
![svgviewer-output_1](https://user-images.githubusercontent.com/30048408/147740261-c05d4e26-387a-4310-9ada-f50bbd6d73a2.png)
![svgviewer-output_2](https://user-images.githubusercontent.com/30048408/147740262-11e3b42c-e91f-4d2b-8605-66362089aa0a.png)
![svgviewer-output_3](https://user-images.githubusercontent.com/30048408/147740263-72313013-9b41-4102-a0cf-97de1d6a23c6.png)
</details>

Reply:
I must have missed those files, it's probably best to use that as a static preview than starting the widgets on the draggable

Reply:
Yes, I think we'll compromise and use the not-so-pretty-but-reliable s_preview system. We recently decided to always call onBuilt before the first widget start... adding onBuilt while the snippet is not in the DOM is not possible so this is risky and too bug prone. I'll change this.

Reply:
In case you ever need to regenerate the images in SVG (I generated those in SVG then changed it to PNG), here is the diff I made, with using [canvas 2 svg lib](https://raw.githubusercontent.com/gliffy/canvas2svg/master/canvas2svg.js):

<details>
<summary>Diff:</summary>

```diff
diff --git addons/website/static/src/snippets/s_countdown/000.js addons/website/static/src/snippets/s_countdown/000.js
index 10074ca17525..5cc34bf95644 100644
--- addons/website/static/src/snippets/s_countdown/000.js
+++ addons/website/static/src/snippets/s_countdown/000.js
@@ -19,6 +19,7 @@ const CountdownWidget = publicWidget.Widget.extend({
      * @override
      */
     start: function () {
+        this.rdectx = new C2S(500,500);
         this.$wrapper = this.$('.s_countdown_canvas_wrapper');
         this.$wrapper.addClass('d-flex justify-content-center');
         this.hereBeforeTimerEnds = false;
@@ -44,6 +45,7 @@ const CountdownWidget = publicWidget.Widget.extend({
         this._initTimeDiff();
 
         this._render();
+        debugger;
 
         this.setInterval = setInterval(this._render.bind(this), 1000);
         return this._super(...arguments);
@@ -206,9 +208,9 @@ const CountdownWidget = publicWidget.Widget.extend({
             for (const val of this.diff) {
                 const canvas = val.canvas.querySelector('canvas');
                 const ctx = canvas.getContext("2d");
-                ctx.canvas.width = this.width;
-                ctx.canvas.height = this.size;
-                this._clearCanvas(ctx);
+                this.rdectx.canvas.width = this.width;
+                this.rdectx.canvas.height = this.size;
+                this._clearCanvas(this.rdectx);
 
                 $(canvas).toggleClass('d-none', hideCountdown);
                 if (hideCountdown) {
@@ -217,14 +219,14 @@ const CountdownWidget = publicWidget.Widget.extend({
 
                 // Draw canvas elements
                 if (this.layoutBackground !== 'none') {
-                    this._drawBgShape(ctx, this.layoutBackground === 'plain');
+                    this._drawBgShape(this.rdectx, this.layoutBackground === 'plain');
                 }
                 this._drawText(canvas, val.nb, val.label, this.layoutBackground === 'plain');
                 if (this.progressBarStyle === 'surrounded') {
-                    this._drawProgressBarBg(ctx, this.progressBarWeight === 'thin');
+                    this._drawProgressBarBg(this.rdectx, this.progressBarWeight === 'thin');
                 }
                 if (this.progressBarStyle !== 'none') {
-                    this._drawProgressBar(ctx, val.nb, val.total, this.progressBarWeight === 'thin');
+                    this._drawProgressBar(this.rdectx, val.nb, val.total, this.progressBarWeight === 'thin');
                 }
                 this.$('.s_countdown_canvas_flex').toggleClass('mx-1', this.layout === 'boxes');
             }
@@ -267,10 +269,10 @@ const CountdownWidget = publicWidget.Widget.extend({
      * Erases the canvas.
      *
      * @private
-     * @param {RenderingContext} ctx - Context of the canvas
+     * @param {RenderingContext} this.rdectx - Context of the canvas
      */
     _clearCanvas: function (ctx) {
-        ctx.clearRect(0, 0, this.size, this.size);
+        this.rdectx.clearRect(0, 0, this.size, this.size);
     },
     /**
      * Draws a text into the canvas.
@@ -284,37 +286,37 @@ const CountdownWidget = publicWidget.Widget.extend({
     _drawText: function (canvas, textNb, textUnit, full = false) {
         const ctx = canvas.getContext("2d");
         const nbSize = this.size / 4;
-        ctx.font = `${nbSize}px Arial`;
-        ctx.fillStyle = this.textColor;
-        ctx.textAlign = 'center';
-        ctx.textBaseline = 'middle';
-        ctx.fillText(textNb, canvas.width / 2, canvas.height / 2);
+        this.rdectx.font = `${nbSize}px Arial`;
+        this.rdectx.fillStyle = this.textColor;
+        this.rdectx.textAlign = 'center';
+        this.rdectx.textBaseline = 'middle';
+        this.rdectx.fillText(textNb, canvas.width / 2, canvas.height / 2);
 
         const unitSize = this.size / 12;
-        ctx.font = `${unitSize}px Arial`;
-        ctx.fillText(textUnit, canvas.width / 2, canvas.height / 2 + nbSize / 1.5, this.width);
+        this.rdectx.font = `${unitSize}px Arial`;
+        this.rdectx.fillText(textUnit, canvas.width / 2, canvas.height / 2 + nbSize / 1.5, this.width);
 
         if (this.layout === 'boxes' && this.layoutBackground !== 'none' && this.progressBarStyle === 'none') {
             let barWidth = this.size / (this.progressBarWeight === 'thin' ? 31 : 10);
             if (full) {
                 barWidth = 0;
             }
-            ctx.beginPath();
-            ctx.moveTo(barWidth, this.size / 2);
-            ctx.lineTo(this.width - barWidth, this.size / 2);
-            ctx.stroke();
+            this.rdectx.beginPath();
+            this.rdectx.moveTo(barWidth, this.size / 2);
+            this.rdectx.lineTo(this.width - barWidth, this.size / 2);
+            this.rdectx.stroke();
         }
     },
     /**
      * Draws a plain shape into the canvas.
      *
      * @private
-     * @param {RenderingContext} ctx - Context of the canvas
+     * @param {RenderingContext} this.rdectx - Context of the canvas
      * @param {boolean} full - if true, the shape will be drawn up to the progressbar
      */
     _drawBgShape: function (ctx, full = false) {
-        ctx.fillStyle = this.layoutBackgroundColor;
-        ctx.beginPath();
+        this.rdectx.fillStyle = this.layoutBackgroundColor;
+        this.rdectx.beginPath();
         if (this.layout === 'circle') {
             let rayon = this.size / 2;
             if (this.progressBarWeight === 'thin') {
@@ -322,61 +324,63 @@ const CountdownWidget = publicWidget.Widget.extend({
             } else {
                 rayon -= full ? 0 : this.size / 10;
             }
-            ctx.arc(this.size / 2, this.size / 2, rayon, 0, Math.PI * 2);
-            ctx.fill();
+            this.rdectx.arc(this.size / 2, this.size / 2, rayon, 0, Math.PI * 2);
+            this.rdectx.fill();
         } else if (this.layout === 'boxes') {
             let barWidth = this.size / (this.progressBarWeight === 'thin' ? 31 : 10);
             if (full) {
                 barWidth = 0;
             }
 
-            ctx.fillStyle = this.layoutBackgroundColor;
-            ctx.rect(barWidth, barWidth, this.width - barWidth * 2, this.size - barWidth * 2);
-            ctx.fill();
+            this.rdectx.fillStyle = this.layoutBackgroundColor;
+            this.rdectx.rect(barWidth, barWidth, this.width - barWidth * 2, this.size - barWidth * 2);
+            this.rdectx.fill();
 
-            const gradient = ctx.createLinearGradient(0, this.width, 0, 0);
+            const gradient = this.rdectx.createLinearGradient(0, this.width, 0, 0);
             gradient.addColorStop(0, '#ffffff24');
             gradient.addColorStop(1, this.layoutBackgroundColor);
-            ctx.fillStyle = gradient;
-            ctx.rect(barWidth, barWidth, this.width - barWidth * 2, this.size - barWidth * 2);
-            ctx.fill();
-            $(ctx.canvas).css({'border-radius': '8px'});
+            this.rdectx.fillStyle = gradient;
+            this.rdectx.rect(barWidth, barWidth, this.width - barWidth * 2, this.size - barWidth * 2);
+            this.rdectx.fill();
+            $(this.rdectx.canvas).css({'border-radius': '8px'});
         }
     },
     /**
      * Draws a progress bar around the countdown shape.
      *
      * @private
-     * @param {RenderingContext} ctx - Context of the canvas
+     * @param {RenderingContext} this.rdectx - Context of the canvas
      * @param {string} nbUnit - how many unit should fill progress bar
      * @param {string} totalUnit - number of unit to do a complete progress bar
      * @param {boolean} thinLine - if true, the progress bar will be thiner
      */
     _drawProgressBar: function (ctx, nbUnit, totalUnit, thinLine) {
-        ctx.strokeStyle = this.progressBarColor;
-        ctx.lineWidth = thinLine ? this.size / 35 : this.size / 10;
+        nbUnit = 21;
+        totalUnit = 100;
+        this.rdectx.strokeStyle = this.progressBarColor;
+        this.rdectx.lineWidth = thinLine ? this.size / 35 : this.size / 10;
         if (this.layout === 'circle') {
-            ctx.beginPath();
-            ctx.arc(this.size / 2, this.size / 2, this.size / 2 - this.size / 20, Math.PI / -2, (Math.PI * 2) * (nbUnit / totalUnit) + (Math.PI / -2));
-            ctx.stroke();
+            this.rdectx.beginPath();
+            this.rdectx.arc(this.size / 2, this.size / 2, this.size / 2 - this.size / 20, Math.PI / -2, (Math.PI * 2) * (nbUnit / totalUnit) + (Math.PI / -2));
+            this.rdectx.stroke();
         } else if (this.layout === 'boxes') {
-            ctx.lineWidth *= 2;
-            let pc = nbUnit / totalUnit * 100;
+            this.rdectx.lineWidth *= 2;
+            let pc = 21;//nbUnit / totalUnit * 100;
 
             // Lines: Top(x1,y1,x2,y2) Right(x1,y1,x2,y2) Bottom(x1,y1,x2,y2) Left(x1,y1,x2,y2)
             const linesCoordFuncs = [
-                (linePc) => [0 + ctx.lineWidth / 2, 0, (this.width - ctx.lineWidth / 2) * linePc / 25 + ctx.lineWidth / 2, 0],
-                (linePc) => [this.width, 0 + ctx.lineWidth / 2, this.width, (this.size - ctx.lineWidth / 2) * linePc / 25 + ctx.lineWidth / 2],
-                (linePc) => [this.width - ((this.width - ctx.lineWidth / 2) * linePc / 25) - ctx.lineWidth / 2, this.size, this.width - ctx.lineWidth / 2, this.size],
-                (linePc) => [0, this.size - ((this.size - ctx.lineWidth / 2) * linePc / 25) - ctx.lineWidth / 2, 0, this.size - ctx.lineWidth / 2],
+                (linePc) => [0 + this.rdectx.lineWidth / 2, 0, (this.width - this.rdectx.lineWidth / 2) * linePc / 25 + this.rdectx.lineWidth / 2, 0],
+                (linePc) => [this.width, 0 + this.rdectx.lineWidth / 2, this.width, (this.size - this.rdectx.lineWidth / 2) * linePc / 25 + this.rdectx.lineWidth / 2],
+                (linePc) => [this.width - ((this.width - this.rdectx.lineWidth / 2) * linePc / 25) - this.rdectx.lineWidth / 2, this.size, this.width - this.rdectx.lineWidth / 2, this.size],
+                (linePc) => [0, this.size - ((this.size - this.rdectx.lineWidth / 2) * linePc / 25) - this.rdectx.lineWidth / 2, 0, this.size - this.rdectx.lineWidth / 2],
             ];
             while (pc > 0 && linesCoordFuncs.length) {
                 const linePc = Math.min(pc, 25);
                 const lineCoord = (linesCoordFuncs.shift())(linePc);
-                ctx.beginPath();
-                ctx.moveTo(lineCoord[0], lineCoord[1]);
-                ctx.lineTo(lineCoord[2], lineCoord[3]);
-                ctx.stroke();
+                this.rdectx.beginPath();
+                this.rdectx.moveTo(lineCoord[0], lineCoord[1]);
+                this.rdectx.lineTo(lineCoord[2], lineCoord[3]);
+                this.rdectx.stroke();
                 pc -= linePc;
             }
         }
@@ -385,36 +389,36 @@ const CountdownWidget = publicWidget.Widget.extend({
      * Draws a full lighter background progressbar around the shape.
      *
      * @private
-     * @param {RenderingContext} ctx - Context of the canvas
+     * @param {RenderingContext} this.rdectx - Context of the canvas
      * @param {boolean} thinLine - if true, the progress bar will be thiner
      */
     _drawProgressBarBg: function (ctx, thinLine) {
-        ctx.strokeStyle = this.progressBarColor;
-        ctx.globalAlpha = 0.2;
-        ctx.lineWidth = thinLine ? this.size / 35 : this.size / 10;
+        this.rdectx.strokeStyle = this.progressBarColor;
+        this.rdectx.globalAlpha = 0.2;
+        this.rdectx.lineWidth = thinLine ? this.size / 35 : this.size / 10;
         if (this.layout === 'circle') {
-            ctx.beginPath();
-            ctx.arc(this.size / 2, this.size / 2, this.size / 2 - this.size / 20, 0, Math.PI * 2);
-            ctx.stroke();
+            this.rdectx.beginPath();
+            this.rdectx.arc(this.size / 2, this.size / 2, this.size / 2 - this.size / 20, 0, Math.PI * 2);
+            this.rdectx.stroke();
         } else if (this.layout === 'boxes') {
-            ctx.lineWidth *= 2;
+            this.rdectx.lineWidth *= 2;
 
             // Lines: Top(x1,y1,x2,y2) Right(x1,y1,x2,y2) Bottom(x1,y1,x2,y2) Left(x1,y1,x2,y2)
             const points = [
-                [0 + ctx.lineWidth / 2, 0, this.width, 0],
-                [this.width, 0 + ctx.lineWidth / 2, this.width, this.size],
-                [0, this.size, this.width - ctx.lineWidth / 2, this.size],
-                [0, 0, 0, this.size - ctx.lineWidth / 2],
+                [0 + this.rdectx.lineWidth / 2, 0, this.width, 0],
+                [this.width, 0 + this.rdectx.lineWidth / 2, this.width, this.size],
+                [0, this.size, this.width - this.rdectx.lineWidth / 2, this.size],
+                [0, 0, 0, this.size - this.rdectx.lineWidth / 2],
             ];
             while (points.length) {
                 const point = points.shift();
-                ctx.beginPath();
-                ctx.moveTo(point[0], point[1]);
-                ctx.lineTo(point[2], point[3]);
-                ctx.stroke();
+                this.rdectx.beginPath();
+                this.rdectx.moveTo(point[0], point[1]);
+                this.rdectx.lineTo(point[2], point[3]);
+                this.rdectx.stroke();
             }
         }
-        ctx.globalAlpha = 1;
+        this.rdectx.globalAlpha = 1;
     },
 });
 
```

</details>